include/rmm/device_buffer.hpp

```diff
@@ -269,13 +300,10 @@ class device_buffer {
     if (new_size <= capacity()) {
       _size = new_size;
     } else {
-      void* const new_data = _mr->allocate(new_size, this->stream());
+      auto tmp = device_buffer{new_size, stream, _mr};
       RMM_CUDA_TRY(
-        cudaMemcpyAsync(new_data, data(), size(), cudaMemcpyDefault, this->stream().value()));
-      deallocate_async();
-      _data     = new_data;
-      _size     = new_size;
-      _capacity = new_size;
+        cudaMemcpyAsync(tmp.data(), data(), size(), cudaMemcpyDefault, this->stream().value()));
+      std::swap(tmp, *this);
```

Comment:
I could have sworn I had a reason for not using CAS here originally, but now I'm struggling to remember what it could have been.

Oh well ¯\_(ツ)_/¯

Reply:
Does this work correctly without adding/specializing a swap function for `device_buffer`?

Reply:
I also have this question. From what I can tell `std::swap` should be safe, but I may be missing something. Also is this "CAS"? There's no comparison, just a swap, right?

Reply:
> Also is this "CAS"?

Sorry, overloaded acronym. "copy and swap"

> Does this work correctly without adding/specializing a swap function for device_buffer?

Yeah, the default implementation of `swap` will just use the move ctor: https://stackoverflow.com/a/25286610/11341974

Reply:
So I read the implementation of swap, and that all seems fine. But could we just move the temporary buffer to this instance instead of swapping?

Reply:
Should be fine as well. No need to move around the empty buffer ;)